test/exe/outofproc_main_test.py

```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/python
+
+import os
+import sys
+import re
+import time
+import tempfile
+import signal
+import subprocess
+import unittest
+from rules_python.python.runfiles import runfiles
+
+class OutOfProcEnvoyTests(unittest.TestCase):
+  """
+  We test that Envoy process behaves correctly.
+  """
+
+  @classmethod
+  def setUp(self):
+    r = runfiles.Create()
+    envoy_path = None
+    if sys.platform == 'win32':
+      envoy_path = r.Rlocation("envoy/source/exe/envoy-static.exe")
+    else:
+      envoy_path = r.Rlocation("envoy/source/exe/envoy-static")
+    if not os.path.isfile(envoy_path):
+      raise Exception("Could not find Envoy")
+
+    envoy_config = r.Rlocation('envoy/test/config/integration/google_com_proxy_port_0.yaml')
+    with open(envoy_config, 'r') as f:
+      content = f.read()
+
+    content = content.replace("{{ ip_any_address }}", "127.0.0.1")
+    content = content.replace("{{ dns_lookup_family }}", "V4_ONLY")
+
+    if sys.platform == 'win32':
+      content = content.replace("{{ null_device_path }}", "NUL")
+    else:
+      content = content.replace("{{ null_device_path }}", "/dev/null")
+
+    envoy_config_new = os.path.join(os.getenv('TEST_TMPDIR'), 'google_com_proxy_port_0.yaml')
+    with open(envoy_config_new, "w") as f:
+      f.write(content)
+
+    try:  # Windows-specific.
+      creationflags = subprocess.CREATE_NEW_PROCESS_GROUP
+    except AttributeError:
+      creationflags = 0
+    self.envoy_proc = subprocess.Popen([envoy_path, "--config-path", envoy_config_new],
+                                       stdin=subprocess.PIPE,
+                                       creationflags=creationflags)
+    time.sleep(2)
+
+  @classmethod
+  def tearDown(self):
+    self.envoy_proc.kill()
+
+  def test_envoy_closes_on_ctrl_break_event(self):
+    self.assertTrue(self.envoy_proc.poll() == None)
+    ctrl_break_signal = None
+    if sys.platform == 'win32':
+      # On Windows sending a ctrl + c signal is quite tricky
+      # so we test that Envoy stops on ctrl + break.
+      ctrl_break_signal = signal.CTRL_BREAK_EVENT
+    else:
+      ctrl_break_signal = signal.SIGTERM
+    self.envoy_proc.send_signal(ctrl_break_signal)
+    time.sleep(2)
+    poll = self.envoy_proc.poll()
+    self.assertTrue(poll != None)
```

Comment:
The assertion that we wait for a bit and the process has exited seems a little loose, if we were to remove the CTRL-BREAK handling logic wouldn't this test still pass?

Could we do some more assertions around maybe synchronizing the process exit or checking for some log line or stat or something that indicates graceful shutdown was invoked via signal? Some similar tests use a test main (instead of the real one) to do things like this, since we have log lines already in the dispatcher callback this might be doable without that?

Reply:
> The assertion that we wait for a bit and the process has exited seems a little loose

Agreed, we should determine our process has finished and then return, not some "pause 30 seconds" hack.

My understanding, ctrl+break/ctrl+close/ctrl+c do not have to be honored, so the timing of the return from those ctrl requests doesn't matter.

Shutdown/logout is time sensitive and manditory, so a process kill follows the return from the ctrl event. This behavior may have evolved since I last implemented it.

Reply:
> Some similar tests use a test main (instead of the real one) to do things like this, since we have log lines already in the dispatcher callback this might be doable without that?

I can parse the standard output to verify that the event was indeed caught and processed.

Reply:
In terms of testing I did the following:

1. I added a "hacky" sleep statement that is responsible for waiting while Envoy is constructed. The alternative to that would be to keep pinging the Envoy ip/port and see when it responds back to know that it is fully initialized.
2. Wait for Envoy to close after we send the singal.
3. Read the standard output to verify that we closed with `ENVOY_WIN32_SIGTERM`

The test needs to be executed outside of bazel as a standalone executable because of the way the console signals work. Bazel is handling them for us and as a result the envoy process never sees them.